### PR TITLE
Algolia Delta Sync

### PR DESCRIPTION
**Pull request by erikkgagewhitt**
### What does this PR do, or why is it needed?

Adds a delta sync queue/job in order to reduce the weight of our Algolia syncs. This will allow us to sync only items that have been created/modified after a certain date. This should drastically reduce our Algolia syncing.

Previously we did a full-sync every day which also included removing every record first. That job has been moved to once a week. The new delta sync will run daily and should have a comparatively minimal amount of records to update.

### What design trade-offs/decisions were made?

I added a new cursor to the ContentItem data source. It functions similarly to byActive, but only fetches after a given date. The date is parsed from a queue that gets passed in, so it can't actually be used in any other situation.

I'm thinking that maybe I should do the date parsing earlier and then just pass the date down instead of the entire queue.

### How do I test this PR?

1. Perform a full sync by uncommenting ln 40 of `/apollos-data-connector-algolia/src/jobs.js`
2. Create a new piece of published content in ROCK
3. Perform a Delta Sync by recommenting ln 40 and uncommenting ln 41 of `/apollos-data-connector-algolia/src/jobs.js` ; the new piece of content should be searchable
5. Update the content item (I like to change the title)
6. Perform another delta sync; the content should be searchable and updated

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [x] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by erikkgagewhitt** [comment]
@vinnyjth This is actually working, for the most part; there are a few things I'm unsure of and deleted items are a big problem. If you have time today to go over it, that would be very helpful.

File: packages/apollos-data-connector-algolia-search/src/data-source.js [hunk 1]
```diff
@@ -31,6 +34,17 @@ export default class Search {
     });
   }
 
+  async updateObjects(args) {
```
Comment (vinnyjth) [inline packages/apollos-data-connector-algolia-search/src/data-source.js]: Can ditch this method and use `addObjects`, same functionality as `save`

File: packages/apollos-data-connector-algolia-search/src/data-source.js [hunk 2]
```diff
@@ -56,6 +70,53 @@ query getItem {
...
+      const items = result.map(({ node }) => node);
+      itemsLeft = items.length === 100;
+
+      if (itemsLeft) args.after = result[result.length - 1].cursor;
+
+      const newlyCreated = [];
+      const newlyModified = [];
+
+      // eslint-disable-next-line
+      for (const item of items) {
```
Comment (vinnyjth) [inline packages/apollos-data-connector-algolia-search/src/data-source.js]: I think you'll be able to remove this logic, if you are just using `addObject`

File: packages/apollos-data-connector-algolia-search/src/jobs.js [hunk 3]
```diff
@@ -1,14 +1,22 @@
...
     return context.dataSources.Search.indexAll();
   });
 
+  DeltaIndexQueue.process(async () => {
+    const context = getContext();
+    return context.dataSources.Search.deltaIndex();
+  });
+
   FullIndexQueue.add(null, { repeat: { cron: '15 3 * * *' } });
+  DeltaIndexQueue.add(null, { repeat: { cron: '15 3 * * 1' } });
```
Comment (vinnyjth) [inline packages/apollos-data-connector-algolia-search/src/jobs.js]: Let's run the `delta` once a day, and teh full index once a week

File: packages/apollos-data-connector-rock/src/content-items/data-source.js [hunk 4]
```diff
@@ -415,6 +415,26 @@ export default class ContentItem extends RockApolloDataSource {
       .cache({ ttl: 60 })
       .andFilter(this.LIVE_CONTENT());
 
+  byDateAndActive = () => {
+    const now = new Date();
+    const datetime = moment(now)
```
Comment (vinnyjth) [inline packages/apollos-data-connector-rock/src/content-items/data-source.js]: Can we use the date that represents the last time the job ran?

File: packages/apollos-data-connector-algolia-search/src/jobs.js [hunk 5]
```diff
@@ -1,14 +1,44 @@
...
+      .map((j) => j.opts.timestamp)
+      .sort((a, b) => {
+        if (a > b) {
+          return -1;
+        }
+        if (a < b) {
+          return 1;
+        }
+        return 0;
+      })[0];
```
Comment (vinnyjth) [inline packages/apollos-data-connector-algolia-search/src/jobs.js]: Does this work safely if there are 0 jobs in the queue?

File: packages/apollos-data-connector-rock/src/content-items/data-source.js [hunk 6]
```diff
@@ -415,6 +415,19 @@ export default class ContentItem extends RockApolloDataSource {
       .cache({ ttl: 60 })
       .andFilter(this.LIVE_CONTENT());
 
+  byDateAndActive = async ({ datetime }) =>
```
Comment (vinnyjth) [inline packages/apollos-data-connector-rock/src/content-items/data-source.js]: Looking good!

File: packages/apollos-data-connector-algolia-search/src/jobs.js (hunk 5 above)
Comment (erikkgagewhitt) [reply packages/apollos-data-connector-algolia-search/src/jobs.js]: Oh, good call. I think that the answer is very likely "no."

I'll have to look a little closer.

File: packages/apollos-data-connector-algolia-search/src/jobs.js [hunk 7]
```diff
@@ -1,14 +1,44 @@
...
+    const date = new Date(timestamp);
+    const datetime = moment(date)
+      .tz(ROCK.TIMEZONE)
+      .format()
+      .split(/[-+]\d+:\d+/)[0];
+    return context.dataSources.Search.deltaIndex({ datetime });
+  });
+
+  FullIndexQueue.add(null, { repeat: { cron: '15 3 * * 1' } });
```
Comment (vinnyjth) [inline packages/apollos-data-connector-algolia-search/src/jobs.js]: Nice! 

Sidenote: I hate how hard cron is to parse at a glance :P thankfully it works so consistently :)

File: packages/apollos-data-connector-algolia-search/src/jobs.js [hunk 8]
```diff
@@ -1,14 +1,44 @@
...
 
   FullIndexQueue.process(async () => {
     const context = getContext();
     return context.dataSources.Search.indexAll();
   });
 
-  FullIndexQueue.add(null, { repeat: { cron: '15 3 * * *' } });
+  DeltaIndexQueue.process(async () => {
+    const context = getContext();
+    const jobs = await DeltaIndexQueue.getCompleted();
```
Comment (vinnyjth) [inline packages/apollos-data-connector-algolia-search/src/jobs.js]: 🙌

File: packages/apollos-data-connector-algolia-search/src/jobs.js (hunk 7 above)
Comment (erikkgagewhitt) [reply packages/apollos-data-connector-algolia-search/src/jobs.js]: Actually, I had a question about this. Is it okay that these are set to run on Monday? Would there be another day that would be better?

Comment (vinnyjth) [reply packages/apollos-data-connector-algolia-search/src/jobs.js]: Monday is actually probably the best date. It's after the initial rush of the weekend and should help clean things up quickly  if content was added / removed during the weekend.

**Comment by erikkgagewhitt** [comment]
@vinnyjth Pushed up the change that handles 0 completed jobs; this should be ready for review again.

**Pull request merged**